src/core/reaction_methods/ReactionAlgorithm.cpp

```diff
@@ -564,38 +564,49 @@ ReactionAlgorithm::generate_new_particle_positions(int type, int n_particles) {
       p_id = get_random_p_id(type, random_index);
     }
     drawn_pids.emplace_back(p_id);
-    // store original position
+    // store original state
     auto const &p = get_particle_data(p_id);
-    old_positions.emplace_back(std::pair<int, Utils::Vector3d>{p_id, p.pos()});
-    // write new position
+    old_state.emplace_back(std::tuple<int, Utils::Vector3d, Utils::Vector3d>{
+        p_id, p.pos(), p.v()});
+    // write new position and new velocity
     auto const prefactor = std::sqrt(kT / p.mass());
     auto const new_pos = get_random_position_in_box();
     move_particle(p_id, new_pos, prefactor);
     check_exclusion_range(p_id);
+    if (particle_inside_exclusion_range_touched) {
+      break;
+    }
   }
 
-  return old_positions;
+  return old_state;
 }
 
 /**
- * Performs a global MC move for a particle of the provided type.
+ * Performs a global MC move for particles of a given type.
+ * Particles are selected without replacement.
+ * @param type     Type of particles to move.
+ * @param n_part   Number of particles of that type to move.
+ * @returns true if all moves were accepted.
  */
-bool ReactionAlgorithm::do_global_mc_move_for_particles_of_type(
-    int type, int particle_number_of_type_to_be_changed) {
+bool ReactionAlgorithm::displacement_move_for_particles_of_type(int type,
+                                                                int n_part) {
   m_tried_configurational_MC_moves += 1;
   particle_inside_exclusion_range_touched = false;
 
-  auto const particle_number_of_type = number_of_particles_with_type(type);
-  if (particle_number_of_type == 0 or
-      particle_number_of_type_to_be_changed == 0) {
+  if (n_part == 0) {
+    // reject
+    return false;
+  }
```

Comment:
That's an oversight. Calling this method with 0 particles doesn't make sense, and it isn't meaningful to return true or false because no move was actually attempted. It's even worse if `n_part` is a negative number... I'll fix that.